### PR TITLE
Updated Readme 2.1 

### PR DESCRIPTION
**Pull request by bhagyashree-sarawate**
@raunakkumar ,Shanbhag, PAVAN Subbayya please review 
Added documentation for helm chart upgradation from 2.0.0 to 2.1.0

### CODE REVIEW

File: helm/charts/hpe-csi-driver/README.md [hunk 1]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
 
 Due to the [helm limitation](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations) to not support upgrade of CRDs between different chart versions, helm chart upgrade is not supported.
 
-Our recommendation is to uninstall the existing chart and install the chart with the desired version. CRDs will be preserved between uninstall and install.
```
Comment (raunakkumar) [inline helm/charts/hpe-csi-driver/README.md:73]: We can still keep this section and add an exception for upgrade to 2.1.0?

Comment (datamattsson) [reply helm/charts/hpe-csi-driver/README.md:73]: Yes, keep this.

File: helm/charts/hpe-csi-driver/README.md [hunk 2]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+For the upgradtion of chart from 2.0.0 to 2.1.0 we need to apply Primera3PAR metata CRDs before the uninstallation of driver as mentioned in steps below:
```
[the remark this answers is not in the document]
Comment (datamattsson) [reply helm/charts/hpe-csi-driver/README.md:73]: Add a `## Upgrading 2.0.0 to 2.1.0` heading.

[the remark this answers is not in the document]
Comment (datamattsson) [reply helm/charts/hpe-csi-driver/README.md:73]: Please use spell check when submitting PRs.

Rewrite L73 as follows:
Before version 2.0.0 is uninstalled, the following CRDs needs to be updated.

File: helm/charts/hpe-csi-driver/README.md [hunk 3]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+```
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumeinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumegroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/snapshotgroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpereplicated_deviceinfo_v2_crd.yaml
+```
+For the upgradtion of the chart from 2.0.0 to 2.1.0 with replication information follow the below steps before the uninstalltion of the driver.
```
[the remark this answers is not in the document]
Comment (datamattsson) [reply helm/charts/hpe-csi-driver/README.md:80]: Rewrite L80: 
If there are HPE Alletra 9000, Primera or 3PAR Remote Copy Groups configured on the cluster, submit the following `Job`:

File: helm/charts/hpe-csi-driver/README.md [hunk 4]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+For the upgradtion of chart from 2.0.0 to 2.1.0 we need to apply Primera3PAR metata CRDs before the uninstallation of driver as mentioned in steps below:
+```
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumeinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumegroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/snapshotgroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpereplicated_deviceinfo_v2_crd.yaml
+```
+For the upgradtion of the chart from 2.0.0 to 2.1.0 with replication information follow the below steps before the uninstalltion of the driver.  
+```
+Create the deployment using the below command, which will modify the rcg-info record to the new key RCGCreatedByCSP.
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:82]: Delete.

File: helm/charts/hpe-csi-driver/README.md [hunk 5]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+kubectl create -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/csi-driver/primera3par/convert-rcg-info.yaml
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:83]: This needs to be rewritten as a Job controller: https://kubernetes.io/docs/concepts/workloads/controllers/job/

File: helm/charts/hpe-csi-driver/README.md [hunk 6]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+
+Once the record is modified, delete the above deployment.
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:85]: Wait for the the `Job` to complete:
```
kubectl wait --for=condition=complete --timeout=600s job/convert-rcg-info
```

File: helm/charts/hpe-csi-driver/README.md [hunk 7]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+kubectl delete -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/csi-driver/primera3par/convert-rcg-info.yaml
+
+Apply the below CRDS
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:88]: Next, apply the new CRDs.
```
kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumeinfos_v2_crd.yaml
kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumegroupinfos_v2_crd.yaml
kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/snapshotgroupinfos_v2_crd.yaml
kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpereplicated_deviceinfo_v2_crd.yaml
```

File: helm/charts/hpe-csi-driver/README.md [hunk 8]
```diff
@@ -70,7 +70,27 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/snapshotgroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpereplicated_deviceinfo_v2_crd.yaml
+```
+For the upgradtion of the chart from 2.0.0 to 2.1.0 with replication information follow the below steps before the uninstalltion of the driver.  
+```
+Create the deployment using the below command, which will modify the rcg-info record to the new key RCGCreatedByCSP.
+kubectl create -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/csi-driver/primera3par/convert-rcg-info.yaml
+
+Once the record is modified, delete the above deployment.
+kubectl delete -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/csi-driver/primera3par/convert-rcg-info.yaml
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:86]: Not needed.

**Review by datamattsson — CHANGES_REQUESTED** [review]

File: helm/charts/hpe-csi-driver/README.md (hunk 5 above)
Comment (datamattsson) [reply helm/charts/hpe-csi-driver/README.md:83]: We also need to remove the infinity sleep from the rcg-info binary and rebuild the image.

Comment (datamattsson) [reply helm/charts/hpe-csi-driver/README.md:83]: Moved manifest to its own release directory: https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/rcg-info/v1.0.0/convert-rcg-info.yaml

File: helm/charts/hpe-csi-driver/README.md (hunk 6 above)
Comment (raunakkumar) [reply helm/charts/hpe-csi-driver/README.md:87]: @bhagyashree-sarawate are you planning to update convert-rcg-info into a job?

Comment (bhagyashree-sarawate) [reply helm/charts/hpe-csi-driver/README.md]: Done the job changes,

File: helm/charts/hpe-csi-driver/README.md (hunk 5 above)
Comment (bhagyashree-sarawate) [reply helm/charts/hpe-csi-driver/README.md]: > We also need to remove the infinity sleep from the rcg-info binary and rebuild the image.

Addressed  this in PR 
https://github.hpe.com/hpe/hpe_3par_primera_rcginfo/pull/2

File: yaml/rcg-info/v1.0.0/convert-rcg-info.yaml [hunk 9]
```diff
@@ -1,24 +1,18 @@
...
-apiVersion: apps/v1
+apiVersion: batch/v1
+kind: Job
 metadata:
-  name: primera3par-rcg-info
-  namespace: hpe-storage
-  labels:
-    app: primera3par-rcg-info
+ name: primera3par-rcg-info
+ namespace: hpe-storage
```
Comment (datamattsson) [inline yaml/rcg-info/v1.0.0/convert-rcg-info.yaml:5]: Omit namespace.

File: helm/charts/hpe-csi-driver/README.md [hunk 10]
```diff
@@ -69,9 +69,34 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+## Upgrading 2.0.0 to 2.1.0
+Before version 2.0.0 is uninstalled, the following CRDs needs to be updated.
+```
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumeinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumegroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/snapshotgroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpereplicated_deviceinfo_v2_crd.yaml
+```
+If there are HPE Alletra 9000, Primera or 3PAR Remote Copy Groups configured on the cluster, follow the below steps before uninstallation
+```
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:83]: Only encase commands code blocks, not descriptive comments.

File: helm/charts/hpe-csi-driver/README.md [hunk 11]
```diff
@@ -69,9 +69,34 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+Create the job using the below commands, which will modify the rcg-info record to the new key RCGCreatedByCSP.
+curl -O  https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/rcg-info/v1.0.0/convert-rcg-info.yaml
+Note: If namespace is other than hpe-storage modify the namespace parameter in convert-rcg-info.yaml file.
+kubectl create -f convert-rcg-info.yaml
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:87]: Rewrite this into: 
Change `kubectl` context into the `Namespace` where the HPE CSI Driver is installed. The most common is "hpe-storage".

```
kubectl config set-context --current --namespace=hpe-storage
```

Create the `Job` using the below commands, which will modify the "rcg-info" record to the new key "RCGCreatedByCSP".

```
kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/yaml/rcg-info/v1.0.0/convert-rcg-info.yaml
```

File: helm/charts/hpe-csi-driver/README.md [hunk 12]
```diff
@@ -69,9 +69,34 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+
+Completion of job status can be verified using below command
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:89]: "using the below command."

File: helm/charts/hpe-csi-driver/README.md [hunk 13]
```diff
@@ -69,9 +69,34 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
+kubectl wait --for=condition=complete --timeout=600s job/primera3par-rcg-info -n hpe-storage
+job.batch/primera3par-rcg-info condition met
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:91]: Omit the output and remove "-n hpe-storage"

File: helm/charts/hpe-csi-driver/README.md [hunk 14]
```diff
@@ -69,9 +69,34 @@ helm install my-hpe-csi-driver hpe-storage/hpe-csi-driver -n hpe-storage -f myva
...
 
+## Upgrading 2.0.0 to 2.1.0
+Before version 2.0.0 is uninstalled, the following CRDs needs to be updated.
+```
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumeinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpevolumegroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/snapshotgroupinfos_v2_crd.yaml
+kubectl apply -f https://raw.githubusercontent.com/hpe-storage/co-deployments/master/helm/charts/hpe-csi-driver/crds/hpereplicated_deviceinfo_v2_crd.yaml
+```
+If there are HPE Alletra 9000, Primera or 3PAR Remote Copy Groups configured on the cluster, follow the below steps before uninstallation
```
Comment (datamattsson) [inline helm/charts/hpe-csi-driver/README.md:82]: Add punctuation to "uninstallation."

**Review by datamattsson — CHANGES_REQUESTED** [review]
A few clarifications.

**Review by datamattsson — CHANGES_REQUESTED** [review]
Submitted a PR [here](https://github.com/bhagyashree-sarawate/co-deployments/pull/3) to ensure users won't get ahead of themselves.

**Comment by bhagyashree-sarawate** [comment]
> Submitted a PR [here](https://github.com/bhagyashree-sarawate/co-deployments/pull/3) to ensure users won't get ahead of themselves.

Should I close this PR ?

**Comment by datamattsson** [comment]
> > Submitted a PR [here](https://github.com/bhagyashree-sarawate/co-deployments/pull/3) to ensure users won't get ahead of themselves.
> 
> Should I close this PR ?

If you merge my PR into your private branch you can push the commit into this PR.

**Comment by bhagyashree-sarawate** [comment]
> > > Submitted a PR [here](https://github.com/bhagyashree-sarawate/co-deployments/pull/3) to ensure users won't get ahead of themselves.
> > 
> > 
> > Should I close this PR ?
> 
> If you merge my PR into your private branch you can push the commit into this PR.

Done

**Review by datamattsson — APPROVED** [review]
LGTM

**Comment by bhagyashree-sarawate** [comment]
@Raunak , @datamattsson Please do not merge this PR , till we get confirmation from test team and Pradip

**Comment by bhagyashree-sarawate** [comment]
> 



> 

We can merge this PR if looks good
Got confirmation from test team , we can go ahead with this change

**Pull request merged**